src/PositionManager.sol

```diff
@@ -168,67 +168,56 @@ contract PositionManager is IPositionManager, ERC721Permit, PoolInitializer, Mul
         return abi.encode(currencyDelta);
     }
 
-    function burn(uint256 tokenId, address sender) internal {
-        _requireApprovedOrOwner(tokenId, sender);
+    /// note: this is overloaded with ERC721Permit._burn
+    function _burn(bytes memory params, address sender) internal returns (bytes memory) {
+        (uint256 tokenId, PoolPosition memory poolPos, bytes memory hookData) =
+            abi.decode(params, (uint256, PoolPosition, bytes));
+
+        _beforeModify(poolPos, tokenId, sender);
+        uint256 liquidity = _beforeBurn(poolPos, tokenId);
 
-        // Checks that the full position's liquidity has been removed and all tokens have been collected from tokensOwed.
-        _validateBurn(tokenId);
+        // Can only call modify if there is non zero liquidity.
+        BalanceDelta delta;
+        if (liquidity > 0) (delta,) = _modifyLiquidity(poolPos, -(liquidity.toInt256()), bytes32(tokenId), hookData);
```

Comment:
`_beforeBurn` actually returns a uint128. So you dont need to use safecast because liquidity cannot be larger than uint128. Also this negation can be unchecked.

```
uint256 liquidity = ...
int256 liquidityDelta;
unchecked {
    liquidityDelta = -int256(liquidity);
}
```